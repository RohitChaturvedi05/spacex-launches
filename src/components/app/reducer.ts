import take from 'lodash/fp/take';
import { ILaunch } from '../../types/launch'


export interface IState {
    compare: boolean,
    data: ILaunch[],
    err: Error | null,
    limit: number,
    missionName: string,
    offset: number,
    rocketName: string,
    selected: string[],
}
export interface IAction {
    type: string,
    payload?: any
}

export const initialState: IState = {
    compare: false,
    data: [],
    err: null,
    limit: 5,
    missionName: '',
    offset: 1,
    rocketName: '',
    selected: []
};


const TYPES = {
    SET_COMPARE: 'SET_COMPARE',
    SET_DATA: 'SET_DATA',
    SET_ERROR: 'SET_ERROR',
    SET_LIMIT: 'SET_LIMIT',
    SET_MISSION_NAME: 'SET_MISSION_NAME',
    SET_OFFSET: 'SET_OFFSET',
    SET_ROCKET_NAME: 'SET_ROCKET_NAME',
    SET_SELECTED: 'SET_SELECTED'
}


export const Actions = {
    setCompare: (payload: boolean) => ({ type: TYPES.SET_COMPARE, payload }),
    setData: (payload: ILaunch[]) => ({ type: TYPES.SET_DATA, payload }),
    setError: (payload: Error) => ({ type: TYPES.SET_DATA, payload }),
    setLimit: (payload: number) => ({ type: TYPES.SET_LIMIT, payload }),
    setMissionName: (payload: string) => ({ type: TYPES.SET_MISSION_NAME, payload }),
    setRocketName: (payload: string) => ({ type: TYPES.SET_ROCKET_NAME, payload }),
    setOffset: (payload: number) => ({ type: TYPES.SET_OFFSET, payload }),
    setSelected: (payload: { id: string[], checked: boolean }) => ({ type: TYPES.SET_SELECTED, payload })
}

export function reducer(state: IState, action: IAction) {
    console.log(action.type, action, state)
    switch (action.type) {
        case TYPES.SET_COMPARE:
            return {
                ...state,
                compare: action.payload
            };
        case TYPES.SET_DATA:
            return {
                ...state,
                data: action.payload
            };
        case TYPES.SET_MISSION_NAME:
            return {
                ...state,
                missionName: action.payload
            };
        case TYPES.SET_ROCKET_NAME:
            return {
                ...state,
                rocketName: action.payload
            };
        case TYPES.SET_OFFSET:
            return {
                ...state,
                offset: action.payload
            };
        case TYPES.SET_SELECTED: {
            const { id, checked } = action.payload
            const selected = checked
                ? take(2, [...state.selected, id])
                : take(2, state.selected.filter((rowid: string) => rowid !== id))

            return {
                ...state,
                selected
            }
        }
        default:
            return state
    }
}
