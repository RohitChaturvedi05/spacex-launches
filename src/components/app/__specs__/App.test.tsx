import React from 'react';
import { mount, shallow } from 'enzyme'
import getLaunchesPast from '../../../services/get-launches-past';
import Compare from '../../compare';
import App from '..';
import raw from './fixture/data.json';
import Rows from '../rows';
import { Button } from '@mui/material';
import * as Reducer from '../reducer';


jest.mock('../../../services/get-launches-past', () => jest.fn(() => Promise.resolve([])))

describe('SpaceX-Launch', () => {


  beforeEach(() => {
    getLaunchesPast.mockResolvedValue(raw.data)
  })

  it('should renders app with correct props', () => {
    const wrapper = shallow(<App />).dive()
    expect(wrapper).toMatchSnapshot()
  });

  it('should open compare', () => {

    const wrapper = shallow(<App />)

    // Should not have Compare
    expect(wrapper.find(Compare)).toHaveLength(0)

    const buttons = wrapper.find(Button)

    // Should have rows
    expect(wrapper.find(Rows)).toHaveLength(1)
    const onSelect = wrapper.find(Rows).prop('onSelect')

    // Select Two records
    onSelect && onSelect('98', true)
    onSelect && onSelect('99', true)

    //Click On Compare button
    buttons.at(0).simulate('click', {})
    wrapper.update();

    // Compare page should be open
    expect(wrapper.find(Compare)).toHaveLength(1)

    const onClose = wrapper.find(Compare).prop('onClose')
    onClose && onClose({})
    wrapper.update();

    // Should have rows not compare dialog
    expect(wrapper.find(Compare)).toHaveLength(0)
    expect(wrapper.find(Rows)).toHaveLength(1)

  });

})
