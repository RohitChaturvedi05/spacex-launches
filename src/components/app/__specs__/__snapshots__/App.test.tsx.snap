// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpaceX-Launch should renders app with correct props 1`] = `ShallowWrapper {}`;
